control-plane-operator/controllers/awsprivatelink/awsprivatelink_controller.go

```diff
@@ -633,6 +637,139 @@ func (r *AWSEndpointServiceReconciler) reconcileAWSEndpointService(ctx context.C
 	return nil
 }
 
+func (r *AWSEndpointServiceReconciler) reconcileAWSEndpointSecurityGroup(ctx context.Context, awsEndpointService *hyperv1.AWSEndpointService, hcp *hyperv1.HostedControlPlane) error {
+	var sgID string
+	var sg *ec2.SecurityGroup
+
+	log := ctrl.LoggerFrom(ctx)
+	var err error
+	sg, err = supportawsutil.GetSecurityGroup(r.ec2Client, vpcEndpointSecurityGroupFilter(hcp.Spec.InfraID, awsEndpointService.Name))
+	if err != nil {
+		return err
+	}
+	if sg != nil {
+		sgID = aws.StringValue(sg.GroupId)
+		if awsEndpointService.Status.SecurityGroupID != sgID {
+			log.Info("WARNING: found existing security group, but it doesn't match status id, updating", "id", sgID)
+			awsEndpointService.Status.SecurityGroupID = sgID
+		}
+	}
+	if sgID == "" {
+		var err error
+		if sg, err = r.createSecurityGroup(ctx, awsEndpointService, hcp); err != nil {
+			return err
+		}
+		sgID = aws.StringValue(sg.GroupId)
+		awsEndpointService.Status.SecurityGroupID = sgID
+	}
+	machineCIDRs := make([]string, len(hcp.Spec.Networking.MachineNetwork))
+	for i, mNet := range hcp.Spec.Networking.MachineNetwork {
+		machineCIDRs[i] = mNet.CIDR.String()
+	}
+	ingressPermissions := supportawsutil.VPCEndpointSecurityGroupRules(machineCIDRs, vpcEndpointPort(awsEndpointService))
+	missingPermissions := diffPermissions(sg.IpPermissions, ingressPermissions)
+	if len(missingPermissions) > 0 {
+		if _, err = r.ec2Client.AuthorizeSecurityGroupIngressWithContext(ctx, &ec2.AuthorizeSecurityGroupIngressInput{
+			GroupId:       aws.String(sgID),
+			IpPermissions: ingressPermissions,
+		}); err != nil {
+			if supportawsutil.AWSErrorCode(err) != "InvalidPermission.Duplicate" {
+				return fmt.Errorf("failed to set security group ingress rules, code: %s", supportawsutil.AWSErrorCode(err))
+			}
+			log.Info("WARNING: got duplicate permissions error when setting security group ingress permissions", "sgID", sgID)
+		}
+		log.Info("authorized security group rules", "id", sgID)
+	} else {
+		log.Info("security group already has required permissions", "id", sgID)
+	}
+	return nil
+}
+
+func vpcEndpointPort(awsEndpointService *hyperv1.AWSEndpointService) int64 {
+	switch awsEndpointService.Name {
+	case manifests.KubeAPIServerPrivateService("").Name:
+		return 6443
+	default:
```

Comment:
todo: This could come from config.KASLBSVCPort
and we can have a new one for RouterLBPort